doc/tutorials/space_shooter/step_4.md

```diff
@@ -0,0 +1,190 @@
+# Adding bullets
+
+For this next step we will add a very important feature to any space shooter game, shooting!
+
+Here is how we will implement it: since we already control our space ship by dragging the screen
+with the mouse/fingers, we will make the ship auto shoot when the player stars the dragging and
+stops shooting when the gesture/input has ended.
+
+So let's start, to begin with it, let's first create a `Bullet` component that will represent the
+shoots in the game.
+
+```dart
+class Bullet extends SpriteAnimationComponent
+    with HasGameReference<SpaceShooterGame> {
+  Bullet({
+    super.position,
+  }) : super(size: Vector2(25, 50));
+
+  @override
+  Future<void> onLoad() async {
+    await super.onLoad();
+
+    animation = await game.loadSpriteAnimation(
+      'bullet.png',
+      SpriteAnimationData.sequenced(
+        amount: 4,
+        stepTime: .2,
+        textureSize: Vector2(8, 16),
+      ),
+    );
+
+    width = 25;
+    height = 50;
+    anchor = Anchor.center;
+  }
+}
+```
+
+So far, this does not introduces any new concept, we just created a component, setting
+up its animations attributes.
+
+A `Bullet` behavior is a simple one, it always moves towards to the top of the screen and should
+be removed from the game if they are not visible anymore, so let's add an `update` method to it
+and make it happen:
+
+```dart
+class Bullet extends SpriteAnimationComponent
+    with HasGameReference<SpaceShooterGame> {
+  Bullet({
+    super.position,
+  }) : super(size: Vector2(25, 50));
+
+  @override
+  Future<void> onLoad() async {
+    // Omitted
+  }
+
+  @override
+  void update(double dt) {
+    super.update(dt);
+
+    position.y += dt * -500;
+
+    if (position.y < -height) {
+      removeFromParent();
+    }
+  }
+}
+```
+
+The above code should be straight forward, but lets break it down:
+
+- We add to the bullet's y axis position at a rate of -500 pixels per second. Remember going up
+in the y axis means getting closer to `0` since the top left corner of the screen is `0, 0`.
+- If the y is smaller than the negative value of the bullet's height, means that the component is
+completely off the screen and it can be removed.
+
+Right, we have a `Bullet` class ready now, so lets make start to implement the action of shooting
+now. First thing, let's create two empty methods in the `Player` class, `startShooting` and
+`stopShooting`.
+
+```dart
+class Player extends SpriteAnimationComponent
+    with HasGameReference<SpaceShooterGame> {
+
+  // Rest of implementation omitted
+
+  void startShooting() {
+    // TODO
+  }
+
+  void stopShooting() {
+    // TODO
+  }
+}
+```
+
+And let's hook into those methods from the game class, we will do that by using the `onPanStart`
+and `onPanEnd` methods from the `PanDetector` mixin that we already have been using for the ship
```

Comment:
Wouldn't it be better to use the `DragCallbacks` instead of the pan detector since that will reach the reader how to do this on any component? And then they'll feel familiar when they read about that later.

Reply:
This was originally written down with the `PanDetector`, `DragCallbacks` was not a thing back when we started it haha.

I could do a refactor on all steps to move to `DragCallbacks`. But I would prefer to do so in a separate PR, wdyt?

Reply:
Sounds good!